Add Service: switchbot_refresh_devices

Vinalti opened the pull request:
## Edits
 - **Add `switchbot_list_devices`**
    _Allow user to list the available switchbot devices from API. (name, type, ID...)_

## Context
When executing a command using switchbot API, the ID of the device must be entered. Unfortunately, the devices' ID are only available via the API, and using authentication. That makes it very hard to do it manually.
The purpose is therefore to solve this issue, by retrieving the informations and make it available for the user.

## Result
Once executed the devices are listed in the logs, like so : 
```logs
2023-01-04 00:50:41.667 WARNING (MainThread) [(...)] {'statusCode': 100, 'body': {'deviceList': [{'deviceId': 'D1XXXXXXXXXX', 'deviceName': 'Hub', 'deviceType': 'Hub Mini', 'enableCloudService': False, 'hubDeviceId': ''}, {'deviceId': 'E5XXXXXXXXXX', 'deviceName': 'Thermometer', 'deviceType': 'Meter', 'enableCloudService': True, 'hubDeviceId': 'D1XXXXXXXXXX'}], 'infraredRemoteList': [{'deviceId': '01-2022XXXXXXXX-XXXXXXXX', 'deviceName': 'LG Beamer', 'remoteType': 'Projector', 'hubDeviceId': 'D1XXXXXXXXXX'}, {'deviceId': '01-2022XXXXXXXX-XXXXXXXX', 'deviceName': 'Light', 'remoteType': 'Light', 'hubDeviceId': 'D1XXXXXXXXXX'}]}, 'message': 'success'}
2023-01-04 00:50:41.670 WARNING (MainThread) [(...)]  --- Native devices ---
2023-01-04 00:50:41.673 WARNING (MainThread) [(...)]   Hub [Hub Mini] -> D1XXXXXXXXXX
2023-01-04 00:50:41.676 WARNING (MainThread) [(...)]   Thermometer [Meter] -> E5XXXXXXXXXX
2023-01-04 00:50:41.678 WARNING (MainThread) [(...)]  --- Infrared devices ---
2023-01-04 00:50:41.681 WARNING (MainThread) [(...)]   Beamer [Projector] -> 01-2022XXXXXXXX-XXXXXXXX
2023-01-04 00:50:41.683 WARNING (MainThread) [(...)]   Light [Light] -> 01-2022XXXXXXXX-XXXXXXXX
```
The logs are available inside the logfile `config/homeassistant.log` but should also be visible from the WebUI (TBC).

_Note: even through the information is not an error or a warning, the loglevel is set to 'Warning' to ensure that most logger displays it._

Vinalti committed:
Add function to list registered devices

Vinalti committed:
Cleanup script file

Vinalti committed:
Update README.md

Vinalti committed:
Add Turn On / Turn Off Commands

add turn On / Off

SiriosDev commented:
This PR doesn't look bad, unfortunately, my server is dead at the moment. As soon as the replacement arrives I will test it and give you a result.
I anticipate though that in my opinion it would be better to put the output in something more user friendly, at the moment I am thinking of a file txt or json

Vinalti commented:
I am not very familliar with the syntax for `pyscripts` but do you think that would be possible to store the result as json or yaml for later use ? 
The idea would be to save the list of devices and then when a user call another function, he can have a dropdown list with the different devices listed? (no more copy paste of device id) 

Also I think the users will be able to see the result easily. The steps I had to go through to install hacs, pyscript and this script were more difficult than opening the logfile.

Vinalti commented:
I did some digging and here is what I found:

### yaml generation
That would be my idea:
- In config file we add an `!include pyscript/apps/switchbot/devices.yaml`
- We add the creation of an empty file by the user
- When calling `switchbot_list_devices` we retrieve the data and we add the devices to `devices.yaml` with all informations we get (incl. deviceID). We will need of course to erase the existing entries everytime (in case a device is not avail. anymore, and to avoid duplicate)
- In the scripts we set the `deviceId` not as a string but as an `Entity` ([ref](https://www.home-assistant.io/docs/blueprint/selectors/#entity-selector))

### Creating Device and State
It seems possible to create an entity just by setting it's state ([source](https://community.home-assistant.io/t/pyscript-how-to-create-a-sensor/455653/4))  
We could use this ability to get/set the device_id somewhere.

Example:  
https://github.com/brewston/Home-Assistant-Strava/blob/86d13c63a78ddb8dca96197c9a02ccc348d45d11/strava.py#L47-L49

SiriosDev commented:
yes, you read my mind, although I was thinking about a list of ids with the information of individual devices as a yaml comment, because I would not know how to create an entity from pyscript

Vinalti commented:
I don't know how to create an entity, but I believe I know :
- How to write data to a file 
- How to `!include` this file in the main `configuration.yaml` file

I believe that should work. what do you think ? 

I will do some tests when I have time.

SiriosDev commented:
some news??? 🙂

Vinalti committed:
Transform list_devices into refresh_devices

Create switch entity for each switchbot IR controlled device

Vinalti commented:
Voilà, that should do it ! 😉

Vinalti committed:
fix switchbot_hvac with new device system.

Vinalti commented:
So, the thing is :
We cannot manage files because the script is running in a sandbox.
However the function below can also add an entity, So that's what I use.
```py
state.set(f'{DOMAIN}.{gen_dev_uid(dev)}', value=dev.get(KEY_DEV_ID), new_attributes=dev )
```

Vinalti committed:
Merge branch 'main' into patch-1

Vinalti committed:
edit HASS Selector for service with entity of type switch.

Vinalti committed:
update README

Vinalti committed:
improve yaml functions docs

Vinalti committed:
Merge branch 'full-featured' into patch-1

SiriosDev commented:
all perfect but I have a question, why limit to only IR devices?

SiriosDev commented:
another question, can you explain to me in broad strokes, "the order of execution" of the various functions, regarding the new.
ah, and then I think it would be useful to add a breaking change with today's date, for those who are already using

Vinalti committed:
clean up code

Vinalti committed:
simplify gen_icon workflow

Vinalti commented:
Yes Of course: 

(Note: I removed the logging and comments in the code below so it's clearer, but it still exists in the real code)

So at first, we just retrieve the result from the APi
```py
def switchbot_refresh_devices():
    headers_dict = auth(**pyscript.app_config)
    url = f"https://api.switch-bot.com/v1.1/devices"
    r = requestGetHelper(url, {}, headers_dict)
```
Check if infrared devices exist and that there is no error.:
```py
    infrared = r.json()['body'].get("infraredRemoteList")
    if infrared is None:
      return None
```
Clear existing switchbot devices (from previous refresh).  
That iterate over the states from `DOMAIN` (`switch.*`) and delete the sensors starting with `PREFIX` (`switchbot_remote_`)
```py
    clear_existing()
###
def clear_existing():
  states = state.names(domain=DOMAIN) # Retrieve all states (=entities) in the `switch.` domain
  prefix = f'{DOMAIN}.{PREFIX}'
  for s in states:
    if s[0:len(prefix)] == prefix:    # If the name contains the prefix
      state.delete(s) # << That is deleting the sensor
```
Then I generate the new devices:
```py
for dev in infrared:
        dev['friendly_name'] = gen_dev_name(dev) # << Gives it a nice friendly name
        dev['icon'] = gen_icon(dev) # Generate an Icon depending on the `deviceType`
        #  vv  Create the device with all the values sent by the API as attributes  vv
        state.set(f'{DOMAIN}.{gen_dev_uid(dev)}', value=dev.get(KEY_DEV_ID), new_attributes=dev )
```
For the generators:
```py
def gen_icon(dev):
  ico = 'remote'  # < default value
  # Create an array of correspondance (can be improved in the future)
  icons = {'Projector': 'projector', 'Light': 'lightbulb-on', 'TV':'television', 'Fan':'fan'}
  typ = dev.get(KEY_DEV_TYPE) # < Get the device type
  if typ in icons:  # < If the dev type is known, use that one instead of the default
    ico = icons[typ]
  return 'mdi:'+ico
```
```py 
# Create a friendly name : "Switchbot <Device-Name> [ IR <Device-Type>]"
def gen_dev_name(dev):
  name = ['Switchbot']
  if (dev.get(KEY_DEV_NAME) is not None):
    name.append(dev.get(KEY_DEV_NAME))
  name.append(f"[IR {dev.get(KEY_DEV_TYPE)}]")
  return ' '.join(name)
```
```py
# Used to generate the unique ID for the device in Home assistant
def gen_dev_uid(dev:dict):
  name = dev.get(KEY_DEV_NAME)              # Get the name
  if name is not None:                      # Format it with:
    name = name.lower()                     # - Lowercase 
    name = re.sub(r'[^0-9a-z]+', '_', name) # - replace any non alphanumeric by '_'
    if name not in ['_', '']:               # If name is fine, return it
      return PREFIX+name
  # if it was fully replaced (e.g name was in another alphabet), return the device type + end of the device_id
  return PREFIX + re.sub(r'[^0-9a-z]+', '_', str(dev.get(KEY_DEV_TYPE)).lower())+'_'+str(dev.get(KEY_DEV_ID)[-4:])
```

And last but not least, the helper function to transform the home assistant uid (e.g. `switch.switchbot_remote_light`) into the switchbot Device ID (more complext than it seems): 
```py
@pyscript_executor
def extract_device_id(device, _recursived=0):
  try: # Try to retrieve the device
    uid = state.get(device)
    return uid
  except NameError: # But if the device doesn't exist, because the state was forgotten (timeout, or whatever)
    if _recursived == 0: # And it is the first time
      switchbot_refresh_devices() # We will force refresh the device list.
      return extract_device_id(device, _recursived=_recursived+1) # And retry the same thing.
    else: # But if we already tried it, we will issue a notification
          # because it means that the device was deleted from switchbot (or it is a wrong device UID).
          # That's useful when the service is integrated into scripts or automations, and the device could 
          # disapears at any time, you want to be aware of that.
      msg=f'Warning: impossible to find {device}. [pyscript SwitchBot]'
      service.call('notify', 'persistent_notification', message=msg)
      service.call('notify', 'notify', message=msg)
      return None
```

SiriosDev commented:
ok, perfect, but is there a reason why we only "scan" ir devices?
As for the breaking changes chapter, I'd rather you do it since you've worked on it and know the changes to bring back to the user.

Vinalti committed:
fix extract_device_id (rm @pyscript_executor)

when @pyscript_executor decorator is present:
it returns a coroutine instead of ID

Vinalti commented:
Ah, yes, that's because it is 2 different objects inside the JSON.
```py
 r.json()['body'].get("infraredRemoteList") # that contains only IR Remotes.
# The normal devices would be:
 r.json()['body'].get("deviceList")
# and KEY_DEV_NAME would be 'deviceType'' instead of 'remoteType'  
```
Technically it is possible, but that ask to either group them as the same (aka. translating `DeviceType` to `RemoteType`) or to create 2 use cases. 
I believe the best would be to extend that with another pull requests. What do you think ?

SiriosDev commented:
>I believe the best would be to extend that with another pull requests. What do you think ?

yeye, I agree

When you have a chance, it would be helpful if you could add a chapter to the Readme file to point out the change from <device id> to <entity>, when it's done I will merge it.

Vinalti commented:
Edit: Wait, do we even have a versionning system yet ? 
Well, if we consider the breaking changes, i believe we should go from version 1.1 to version 2.0 ? Maybe adding the keyword `beta` can be a wise choice as to make sure people undestand it's not a fully tested stable release ? (I tested it, but only with my setup, I just have a couple of devices.)

SiriosDev commented:
> Edit: Wait, do we even have a versionning system yet ? Well, if we consider the breaking changes, i believe we should go from version 1.1 to version 2.0 ? Maybe adding the keyword `beta` can be a wise choice as to make sure people undestand it's not a fully tested stable release ? (I tested it, but only with my setup, I just have a couple of devices.)

no we don't have versionig yet, because not using relases I wouldn't know how to insert it, do you know how to do it?

Vinalti commented:
I think we can use tags in the commits. But in a more general way we can just give a number to the different version.
The number can be mentioned in the commit message, and must be updated at every pull request merge in the readme for instance.
Also, when merging, it's good to rebase and squash the commits so the main branch is clean.

And for the version number we can follow this pattern:
`<main>.<feature>.<fix>`
- `<main>`  being the main version, inducing a big change or a breaking change
- `<feature>` is incremented when a new feature is released
- `<fix>` is incremented only when a bug fix or small changes are released.

SiriosDev commented:
> I think we can use tags in the commits. But in a more general way we can just give a number to the different version. The number can be mentioned in the commit message, and must be updated at every pull request merge in the readme for instance. Also, when merging, it's good to rebase and squash the commits so the main branch is clean.
> 
> And for the version number we can follow this pattern: `<main>.<feature>.<fix>`
> 
> * `<main>`  being the main version, inducing a big change or a breaking change
> * `<feature>` is incremented when a new feature is released
> * `<fix>` is incremented only when a bug fix or small changes are released.

I am new to the git world, I hope I understand everything 😅, so I will trust you.
if you agree I would start from 0.1.0 and call the previous one "unnumbered"

Vinalti committed:
Cleanup and refresh ReadMe

More readable formatting
Remove duplicated entries
Add changelog  (incl. version number)

Vinalti commented:
Ok, please have a look to the ReadMe file. I believe that should be good now

SiriosDev commented:
So I have to do this right?
![image](https://user-images.githubusercontent.com/26876994/212479293-e03914d5-9627-4ea0-aeca-59b325722b4f.png)

SiriosDev commented:
as you can see, HA does not recognize the unique ID, when I have time I work on it, but if you come up with something let me know

![image](https://user-images.githubusercontent.com/26876994/212500061-e7a46a54-f53d-4dd9-9409-1c59ec161672.png)

Vinalti commented:
> as you can see, HA does not recognize the unique ID, when I have time I work on it, but if you come up with something let me know
> 

Well, that's normal with custom sensors, I tried adding a `unique_id` according to the [documentation](https://developers.home-assistant.io/docs/entity_registry_index/), that doesn't help, still the same error.
That doesn't prevent the tool from working properly as we don't use that section.

SiriosDev commented:
Ok
